### PR TITLE
Add auto-reload upon file timestamp change

### PR DESCRIPTION
**Pull request by federicotdn**
**Proposed changes**:
- Add a new CLI argument, `--auto-reload`. If enabled, modules containing actions will be reloaded when an HTTP request is received, but only if the files have been changed since they were last loaded.
- Closes https://github.com/RasaHQ/rasa/issues/4058.

Implementation details:

When we load a regular package (directory with `__init__.py` inside), or a file module (e.g. `actions.py`), the file path and last-modified timestamp is recorded. Whenever an HTTP request is received, before running the actions, all current file timestamps are checked against the recorded ones (there is no polling or file modification events involved). This process is quick because it only involves comparing floats. All files that have been modified since the last check are re-loaded using `importlib.reload`. After that, the whole `Action` class hierarchy is re-scanned, and the classes are registered.

One of the problems of re-loading modules is that the classes that have been already defined there aren't deleted or anything, so they are still part of the `Action` hierarchy. The consequence of this is that if a module containing `MyAction` is re-loaded, then there will be two `MyAction` classes in the hierarchy. To get around this, each time an action class is registered, it is "tagged" with `action_class._sdk_loaded = True`. That way, when we re-scan the `Action` class hierarchy, we know what actions we have already loaded in the past, and we can ignore them.

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa-sdk/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa-sdk#code-style) for instructions)

### CODE REVIEW

File: rasa_sdk/executor.py [hunk 1]
```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
...
+            # package itself.
+            timestamp = os.path.getmtime(module_file)
+            self._modules[module_file] = (timestamp, module)
+
+        return module
+
     def register_package(self, package: Union[Text, types.ModuleType]) -> None:
-        from rasa_sdk.interfaces import Action
+        """Register all the `Action` subclasses contained in a Python module or
+        package.
```
Comment (wochinge) [inline rasa_sdk/executor.py:239]: 🚀 Let's document that this function potentially kills everything in case loading the module goes wrong.

File: rasa_sdk/executor.py [hunk 2]
```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
...
         for action in actions:
-            meta = action.__dict__.get("Meta", False)
-            abstract = getattr(meta, "abstract", False)
             if (
                 not action.__module__.startswith("rasa_core.")
                 and not action.__module__.startswith("rasa.")
                 and not action.__module__.startswith("rasa_sdk.")
                 and not action.__module__.startswith("rasa_core_sdk.")
-                and not abstract
+                and not inspect.isabstract(action)
```
Comment (wochinge) [inline rasa_sdk/executor.py:265]: how does this function work? :-D

Comment (wochinge) [reply rasa_sdk/executor.py:265]: don't we need to do `Action(ABC)` then or use `@abstractmethod`??

File: rasa_sdk/executor.py [hunk 3]
```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
...
+        is larger than the previous one, the module is re-imported using
+        `importlib.reload`.
+
+        If one or more modules are reloaded during this process, the entire
+        `Action` class hierarchy is scanned again to see what new classes can
+        be registered.
+        """
+        to_reload = {}
+
+        for path, (timestamp, module) in self._modules.items():
```
Comment (wochinge) [inline rasa_sdk/executor.py:283]: I'd split that method up in `_find_modules_to_reload` and `_reload_modules` and avoid having such a long method body.

File: rasa_sdk/executor.py [hunk 4]
```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
...
+            try:
+                new_timestamp = os.path.getmtime(path)
+            except OSError:
+                # Ignore missing files
+                continue
+
+            if new_timestamp > timestamp:
+                to_reload[path] = (new_timestamp, module)
+
+        module_reloaded = False
```
Comment (wochinge) [inline rasa_sdk/executor.py:293]: ?
```suggestion
        all_modules_reloaded = False
```

File: setup.cfg [hunk 5]
```diff
@@ -13,4 +13,4 @@ license_file = LICENSE.txt
 
 [flake8]
 max-line-length = 88
-ignore = W503, E121, E126, E211, E225, E501, E203, E402, F401, F811, F821
+ignore = W503, E121, E126, E211, E225, E501, E203, E402, F401, F811
```
Comment (wochinge) [inline setup.cfg:16]: Why this change?

File: tests/test_executor.py [hunk 6]
```diff
@@ -1,7 +1,42 @@
...
+        return []
+"""
+
+
+def _write_action_file(
+    package_path: Text,
+    file_name: Text,
+    class_name: Text,
+    action_name: Text,
+    message: Text = None,
```
Comment (wochinge) [inline tests/test_executor.py:31]:
```suggestion
    message: Optional[Text] = None,
```

File: tests/test_executor.py [hunk 7]
```diff
@@ -30,3 +65,135 @@ def test_deprecated_utter_elements():
         "image": None,
         "attachment": None,
     }
+
+
+@pytest.fixture()
+def package_path() -> None:
```
Comment (wochinge) [inline tests/test_executor.py:71]:
```suggestion
def package_path() -> Generator[Text, None, None]:
```

File: rasa_sdk/executor.py [hunk 8]
```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
...
         self, package: Union[Text, types.ModuleType], recursive: bool = True
     ) -> None:
-        """ Import all submodules of a module, recursively, including
-        subpackages
+        """Import a module, or a package and its submodules recursively.
 
-        :param package: package (name or actual module)
-        :type package: str | module
-        :rtype: dict[str, types.ModuleType]
+        Args:
```
Comment (wochinge) [inline rasa_sdk/executor.py:198]: 🚀

File: rasa_sdk/executor.py [hunk 9]
```diff
@@ -176,54 +182,135 @@ def register_function(self, name: Text, f: Callable) -> None:
...
+            except OSError:
+                # Ignore missing files
+                continue
+
+            if new_timestamp > timestamp:
+                to_reload[path] = (new_timestamp, module)
+
+        module_reloaded = False
+
+        for path, (timestamp, module) in to_reload.items():
```
Comment (wochinge) [inline rasa_sdk/executor.py:295]: should we run this in the background using `sanic_app.add_task`?

File: rasa_sdk/executor.py (hunk 2 above)
Comment (federicotdn) [reply rasa_sdk/executor.py:265]: Yes! I changed it because that's the canonical way of making abstract classes in Python. I wasn't able to find why the previous method was being used.

File: rasa_sdk/executor.py (hunk 4 above)
Comment (federicotdn) [reply rasa_sdk/executor.py:293]: `any_modules_reloaded` is better I think, since we only care if at least one was reloaded

File: setup.cfg (hunk 5 above)
Comment (federicotdn) [reply setup.cfg:16]: That's so that we get warnings for using variables that are not defined, not sure why it was disabled since it's super useful.

File: rasa_sdk/executor.py (hunk 9 above)
Comment (federicotdn) [reply rasa_sdk/executor.py:295]: In that case the task would run after the HTTP request has been handled, so that modules would get reloaded after the request. This would demand users to make two requests instead of one to see the new changes, which I think would be unexpected for them.

Comment (federicotdn) [reply rasa_sdk/executor.py:314]: I think, since this feature is for development use only and that checking for timestamps is a quick process (specially with small amounts of files), that reloading them before handling the requests won't be a problem.

**Comment by m-vdb** [comment]
@federicotdn @wochinge I just ran into this "issue" myself and spent 30min trying to understand why my slots where not picked up by my form. I simply assumed that the action server would reload automatically (maybe because I am familiar with Django?).

I am wondering if we wouldn't want to make it the default behaviour (and an option to disable it). Are there drawbacks to that approach?
Maybe we also need to update the doc on that specific aspect: "you need to reload your action server if file A and B change"?

**Comment by federicotdn** [comment]
@m-vdb From what I've seen, auto-reloading of files always seems to be a development-only option (for example in Flask and Sanic). I think it makes sense to have it disabled by default, because in one way or another it always implies that you're doing some additional processing to check if the files have been updated or not (via polling, OS-level file change watching, etc.). I agree that we should update the docs though, maybe we could add also that this new feature exists as well.

**Comment by m-vdb** [comment]
ok makes sense re development, and re docs 👍

File: rasa_sdk/executor.py (hunk 9 above)
Comment (wochinge) [reply rasa_sdk/executor.py:314]: makes sense 👍

File: rasa_sdk/executor.py (hunk 2 above)
Comment (wochinge) [reply rasa_sdk/executor.py:272]: @federicotdn Did you do this one yet?

File: rasa_sdk/endpoint.py [hunk 10]
```diff
@@ -111,9 +130,12 @@ def run(
     ssl_certificate: Optional[Text] = None,
     ssl_keyfile: Optional[Text] = None,
     ssl_password: Optional[Text] = None,
+    auto_reload: bool = False,
 ):
```
Comment (wochinge) [inline rasa_sdk/endpoint.py:134]:
```suggestion
) -> NoReturn:
```

**Comment by federicotdn** [comment]
Since it's for development I think we can leave it. Also I couldn't find a tidy way of unloading modules permanently in Python.

File: rasa_sdk/executor.py (hunk 2 above)
Comment (federicotdn) [reply rasa_sdk/executor.py:272]: What would you change here? The project itself does not have any abstract classes that I could find 🤔

Comment (wochinge) [reply rasa_sdk/executor.py:272]: ah, true. `class Action` is filtered out by the other checks anyway

**Comment by JulianGerhard21** [comment]
@federicotdn @m-vdb 

If I understand the implementation correctly, only the action classes that have already been loaded initially are reloaded, i.e. if I separate action classes into separate files in the "Actions" folder, for example, and make changes to them or add a new file, they are not loaded or reloaded.

Is this correct?
